docs/book/src/migration/guide.md

```diff
@@ -0,0 +1,192 @@
+│           └── zz_generated.deepcopy.go
+├── controller
+└── webhook
+```
+
+In `register.go`, you can find something like
+```go
+SchemeGroupVersion = schema.GroupVersion{Group: {your-group}.{your-domain}, Version: {your-version}}
+```
+
+Now you have figured out what are your domain, group, version and kind.
+
+## Initialize the Project
+
+Run the following command
+
+```bash
+kubebuilder init --domain {your-domain}
+```
+
+## Migrate API and Controller
+
+Run the following command to scaffold the API and the controller. Please follow the prompt to scaffold the API and (or)
+the controller.
+
+```bash
+kubebuilder create api --group {your-group} --version {your-version} --kind {Your-kind}
+```
+
+You can run it multiple times if you have more than one API and controller. If you need multiple groups, please follow
+[this document](../TODO.md)
+
+
+### Migrate the API
+
+Now copy your API definition from `pkg/apis/{your-group}/{your-version}/{your-kind}_types.go` to
+`api/{your-version}/{your-kind}_types.go`. You only need to copy the implementation of your `Spec` and `Status`.
+
+You don't need to copy these markers from v1 project
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+```
+
+You need to ensure marker `// +kubebuilder:object:root=true` exists in v2 project and is the 2nd closest comment for
+your root object. e.g. `CronJob` and `CronJobList`
+
+```go
+// +kubebuilder:object:root=true
+
+// CronJob is the Schema for the cronjobs API
+type CronJob struct {...}
+```
+
+### Migrate the Controller
+
+Now copy your controller implementation from `pkg/controller/{your-kind}/{your-kind}_controller.go`
+to `controllers/{your-kind}_controller.go`.
+
+Copy the fields from your old reconciler struct to the new one.
+
+Copy the business logic code of `Reconcile` function.
+
+Copy the rbac related [markers](../TODO.md) to the new controller file.
+
+Copy the code under `func add(mgr manager.Manager, r reconcile.Reconciler) error` to
+`SetupWithManager(mgr ctrl.Manager) error`.
+
+## Migrate the Webhook
+
+If you don't have a webhook, you can skip this section.
+
+## Core Types and Foreign CRDs
+
+If you are using webhook for k8s core types (e.g. Pods) or for a CRD that is not owned by you, you can refer the
+[controller-runtime example for builtin types](https://github.com/kubernetes-sigs/controller-runtime/tree/master/examples/builtins)
+and do something similar. Kubebuilder doesn't scaffold much for these cases, but you can use the library in
+controller-runtime.
+
+## Scaffold Webhooks
+
+Run the following command to scaffold the webhook for your CRD.
+Specify `--defaulting` and `--programmatic-validation` flags as need.
+
+```bash
+kubebuilder create webhook --group {your-group} --version {your-version} --kind {Your-kind} --defaulting --programmatic-validation
+```
+
+You can run it multiple times if you need multiple webhooks.
+
+### Validating Webhooks
```

Comment:
these headings feel redundant.  just remove these subheads and make this a single section.  For example:

> Now, we'll need to copy the logic for each webhook.  For validating webhooks, we can copy the contents of `func validating{...}Fn` to `ValidateCreate`, and similarly ... for defaulting webhooks.

Reply:
they also seem to break up the flow -- I keep thinking I've hit the end, and finding more, and not knowing if it's part of the next section, or this one.